notes.txt

```diff
@@ -0,0 +1,241 @@
+    enum RecordType {
+      DEPOSIT = 0;
+      REWARDS = 1;
+      WITHDRAWAL = 2;
+    }
+    enum Status {
+      TRANSFER = 0;
+      STAKE = 1;
+    }
+    Status status = 5; 
+  }
+  - Add next status types
+      - transfer: the balance must be transferred to the host zone
+      - stake: the balance must be staked
+# 1 Aggregate into a single deposit record / successful epoch / chain
+# 2 Update DepositRecord type
+# 3 Update Epoch processing code
+
+>>> EPOCH START
+----------
+
+# users call liquid stake
+def liquidStake(tokens)
+  currentRecord = getRecord(epochId)
+  currentRecord.amount += tokens.amount
+
+>>> EPOCH END
+for record in depositRecords:
+  hz = getHz(record)
+  if record.status == TRANSFER:
+    def ibcTransferAck(ack):
+      if ack == success:
+        record = getRecord(ack)
+        delete(record)
+    # add record deletion logic to ack
+    ibcTransfer.ackFunction = ibcTransferAck
+    # ibc transfer the tokens over
+    ibcTransfer(to=hz.depositAccount(), amount=record.amount)
+
+----------
+>>> NEXT EPOCH STARTS
+
+
+
+
+
+
+
+
+
+*******************************************************************
+******************************REWARDS******************************
+*******************************************************************
+
+# SECTION 2
+# 0 Set withdrawal address of delegation account to withdrawal account in host zone
+    registration ack
+# 1 Update DepositRecord to handle rewards types
+# 2 Update epoch end to process all records for each host zone
+# 3 Add new records for rewards that accrued in the previous epoch
+
+>>> EPOCH START
+----------
+... rewards accrue via calls to liquid stake
+
+>>> EPOCH END
+
+# process all unprocessed records for each host zone
+for record in records where record.type == REWARD:
+  def bankSendAck(ack):
+    record = getRecord(ack.id)
+    record.type = DEPOSIT
+    record.status = STAKE
+  # do upon completion
+  ICABankSend.Ack = bankSendAck
+  ICABankSend(from=withdrawal, to=delegat, amount=record.amount)
+
+# add new records for rewards that accrued in the previous epoch
+for hz in getAllHostZones():
+  def addRewardsRecord(hz):
+    # define a query callback function
+    def queryCB(hz):
+      recordBalance = sumAllRewardsRecords()
+      newBalance = balance - recordBalance
+      createNewRewardsRecord(newBalance)
+    wA = hz.withdrawalAccount()
+    query(address=wA, queryType=balance, cb=queryCB)
+
+
+# NOTE: this is executed after processing records because although ICA calls
+# _are_ ordered (meaning we could first send balances from withdrawal -> delegate account),
+# there's no way to enforce that the records are updated on stride without chaining these.
+# unclear whether chaining is a robust approach. What happens if one step fails?
+# what do we have to roll back?
+# easier to think about these in isolation IMO
+1. ICQ wA balance
+2. ICQ CB: create records, BankSend ICA call them from withdrawal -> delegate account
```

Comment:
edge case logic: when you BankSend ICA Call, mark records 'in progress' in case the ICA ack takes longer to return than an epoch (in which case you risk triggering the BankSend ICA Call another time in the next epoch before the first BankSend ICA Call acks).

Reply:
These notes evolved through the commits, in it's current state all timeouts must be l.t. the epoch length